File: Longest Increasing Subsequence - Leetcode 300.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        # Bottom Up DP (Tabulation)
        # Time: O(n^2)
        # Space: O(n)
        
        n = len(nums)
        dp = [1] * n

        for i in range(1, n):
            for j in range(i):
                if nums[i] > nums[j]:
                    dp[i] = max(dp[i], dp[j] + 1)

        return max(dp)

File: Longest Increasing Subsequence - Leetcode 300/Longest Increasing Subsequence - Leetcode 300.cpp
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        int n = nums.size();
        vector<int> dp(n, 1);

        for (int i = 1; i < n; i++) {
            for (int j = 0; j < i; j++) {
                if (nums[i] > nums[j]) {
                    dp[i] = max(dp[i], dp[j] + 1);
                }
            }
        }

        return *max_element(dp.begin(), dp.end());
    }
};


class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        vector<int> lis;
        
        for(auto num: nums){
            
            auto it = lower_bound(lis.begin(),lis.end(),num);
            
            if(it == lis.end()) lis.push_back(num);
            else *it = num;
        
        }
        
        return lis.size();
    }

    // Time: O(nlog(n))
    // Space: O(n)
    // Binary search optimization
};
